gateway/src/core.rs

```diff
@@ -0,0 +1,69 @@
+use crate::p2p::{P2P, P2PChannels};
+use common::{
+    gm_request_response::Request,
+    meta_exchange::{Response, Role},
+};
+pub struct Gateway {
+    p2p_channels: P2PChannels,
+    p2p: Value<P2P>,
+}
+
+enum Value<T> {
+    Here(T),
+    Moved,
+}
+
+impl<T> Value<T> {
+    /// extracts value
+    /// can be performed only once if Value::Here, otherwise it will panic
+    fn extract(&mut self) -> T {
+        let mut moved = Value::Moved;
+        std::mem::swap(&mut moved, self);
+
+        match moved {
+            Value::Here(val) => return val,
+            Value::Moved => panic!("was already moved"),
+        }
+    }
+
+    fn contains(&self) -> bool {
+        match self {
+            Value::Here(_) => true,
+            Value::Moved => false,
+        }
+    }
+}
+
+impl Gateway {
+    pub fn new(node_urls: Vec<String>) -> Result<Gateway, Box<dyn std::error::Error>> {
+        let mut p2p = P2P::new(node_urls)?;
+        // TODO: prepare works in background and you can't start sending requests immediately when you created Gateway
+        // I need to create some sort of state/flags or block the thread that can prevent sending requests before initialization even happened
+        p2p.prepare().map_err(|e| format!("prepare: {}", e))?;
+
+        let p2p_channels = p2p.interface_channels();
+
+        return Ok(Gateway {
```

Comment:
Nit: why `return`, just `Ok(...)` w/o the semicolon. Also, you don't really need to have an extra variable for `p2p_channels`, but that's pure style I think.